streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -166,6 +166,13 @@ private void initialize(final RestoringTasks active) {
                     restorer.setRestoredOffset(0);
                     iter.remove();
                     completedRestorers.add(topicPartition);
+                } else if (restorer.checkpoint() >= restorer.offsetLimit()) {
```

Comment:
Please pay special attention to this.

Reply:
I've added some test cases for this and I think this is OK based on how `StateRestorer` behaves, but still appreciate the scrutiny.

Reply:
Could you merge this condition with the first to `restorer.checkpoint() >= endOffset || restorer.checkpoint() >= restorer.offsetLimit()`? 
Since all branches are very similar, it would also be good to extract
```
restorer.setRestoredOffset(restorer.checkpoint());
iter.remove();
completedRestorers.add(topicPartition);
```
to a private method with a parameter for the argument for `setRestoredOffset()`.